anonymous scopes break integration tests expecting specific scopes

```
--- FAIL: TestTcProxy (10.40s)
    tcproxy_test.go:57: Got incorrect data: {
          "scopes": [
            "assume:anonymous",
            "auth:current-scopes",
            "auth:expand-scopes",
            "auth:get-client:*",
            "auth:get-role:*",
            "auth:list-clients",
            "auth:list-roles",
            "github:get-badge:*",
            "github:get-repository:*",
            "github:latest-status:*",
            "github:list-builds",
            "hooks:get:*",
            "hooks:list-hooks:*",
            "hooks:list-last-fires:*",
            "hooks:status:*",
            "index:find-task:*",
            "index:list-namespaces:*",
            "index:list-tasks:*",
            "purge-cache:all-purge-requests",
            "purge-cache:purge-requests:*",
            "queue:get-artifact:SampleArtifacts/_/X.txt",
            "queue:get-artifact:public/*",
            "queue:get-provisioner:*",
            "queue:get-task:*",
            "queue:get-worker-type:*",
            "queue:get-worker:*",
            "queue:list-artifacts:*",
            "queue:list-dependent-tasks:*",
            "queue:list-provisioners",
            "queue:list-task-group:*",
            "queue:list-worker-types:*",
            "queue:list-workers:*",
            "queue:pending-count:*",
            "queue:status:*",
            "secrets:list-secrets",
            "worker-manager:get-worker-pool:*",
            "worker-manager:get-worker:*",
            "worker-manager:list-providers",
            "worker-manager:list-worker-pool-errors:*",
            "worker-manager:list-worker-pools",
            "worker-manager:list-workers:*"
          ]
        }
FAIL
```

--- a/changelog/issue-3834.md
+++ b/changelog/issue-3834.md
@@ -1,0 +1,4 @@
+audience: general
+level: silent
+reference: issue 3834
+---

--- a/clients/client-go/creds_test.go
+++ b/clients/client-go/creds_test.go
@@ -79,7 +79,7 @@ func Test_PermaCred(t *testing.T) {
 		RequiredScopes: []string{"scope:this"},
 	})
 	checkAuthenticate(t, response, err,
-		"tester", []string{"scope:*"})
+		"tester", []string{"assume:anonymous", "scope:*"})
 }
 
 func Test_TempCred(t *testing.T) {
@@ -94,7 +94,7 @@ func Test_TempCred(t *testing.T) {
 		RequiredScopes: []string{"scope:1"},
 	})
 	checkAuthenticate(t, response, err,
-		"tester", []string{"scope:1", "scope:2"})
+		"tester", []string{"assume:anonymous", "scope:1", "scope:2"})
 }
 
 func Test_NamedTempCred(t *testing.T) {
@@ -110,7 +110,7 @@ func Test_NamedTempCred(t *testing.T) {
 		RequiredScopes: []string{"scope:1"},
 	})
 	checkAuthenticate(t, response, err,
-		"jimmy", []string{"scope:1", "scope:2"})
+		"jimmy", []string{"assume:anonymous", "scope:1", "scope:2"})
 }
 
 func Test_PermaCred_Bewit(t *testing.T) {
@@ -259,7 +259,7 @@ func Test_AuthorizedScopes(t *testing.T) {
 		RequiredScopes: []string{"scope:1"},
 	})
 	checkAuthenticate(t, response, err,
-		"tester", []string{"scope:1", "scope:3"})
+		"tester", []string{"assume:anonymous", "scope:1", "scope:3"})
 }
 
 func Test_TempCredWithAuthorizedScopes(t *testing.T) {
@@ -275,7 +275,7 @@ func Test_TempCredWithAuthorizedScopes(t *testing.T) {
 		RequiredScopes: []string{"scope:1"},
 	})
 	checkAuthenticate(t, response, err,
-		"tester", []string{"scope:1"})
+		"tester", []string{"assume:anonymous", "scope:1"})
 }
 
 func Test_NamedTempCredWithAuthorizedScopes(t *testing.T) {
@@ -292,5 +292,5 @@ func Test_NamedTempCredWithAuthorizedScopes(t *testing.T) {
 		RequiredScopes: []string{"scope:1"},
 	})
 	checkAuthenticate(t, response, err,
-		"julie", []string{"scope:1"})
+		"julie", []string{"assume:anonymous", "scope:1"})
 }

--- a/clients/client-go/tcauth/auth_examples_test.go
+++ b/clients/client-go/tcauth/auth_examples_test.go
@@ -11,16 +11,7 @@ import (
 
 func Example_scopes() {
 
-	// Note: the API call we will make doesn't need credentials as it supplies public information.
-	// However, for the purpose of demonstrating the general case, this is how you can provide
-	// credentials for API calls that require them.
-	myAuth := tcauth.New(
-		&tcclient.Credentials{
-			ClientID:    "SOME-CLIENT-ID",
-			AccessToken: "SOME-WELL-FORMED-ACCESS-TOKEN",
-		},
-		"https://community-tc.services.mozilla.com",
-	)
+	myAuth := tcauth.New(nil, "https://community-tc.services.mozilla.com")
 
 	// Look up client details for client id "project/taskcluster/generic-worker/taskcluster-ci"...
 	resp, err := myAuth.Client("project/taskcluster/generic-worker/taskcluster-ci")

--- a/clients/client-py/test/test_async.py
+++ b/clients/client-py/test/test_async.py
@@ -32,7 +32,7 @@ def test_async_works_with_permanent_credentials():
                 'clientScopes': ['test:a'],
                 'requiredScopes': ['test:a'],
             })
-            assert result == {'scopes': ['test:a'], 'clientId': 'tester'}
+            assert result == {'scopes': ['assume:anonymous', 'test:a'], 'clientId': 'tester'}
 
     loop.run_until_complete(x())
 
@@ -60,6 +60,6 @@ def test_async_works_with_temporary_credentials():
                 'clientScopes': ['test:*'],
                 'requiredScopes': ['test:xyz'],
             })
-            assert result == {'scopes': ['test:xyz'], 'clientId': 'tester'}
+            assert result == {'scopes': ['assume:anonymous', 'test:xyz'], 'clientId': 'tester'}
 
     loop.run_until_complete(x())

--- a/clients/client-py/test/test_client.py
+++ b/clients/client-py/test/test_client.py
@@ -774,7 +774,7 @@ def test_permacred_simple():
         'clientScopes': ['test:a'],
         'requiredScopes': ['test:a'],
     })
-    assert result == {'scopes': ['test:a'], 'clientId': 'tester'}
+    assert result == {'scopes': ['assume:anonymous', 'test:a'], 'clientId': 'tester'}
 
 
 def test_permacred_simple_authorizedScopes():
@@ -790,7 +790,7 @@ def test_permacred_simple_authorizedScopes():
         'clientScopes': ['test:*'],
         'requiredScopes': ['test:a'],
     })
-    assert result == {'scopes': ['test:a', 'test:b'], 'clientId': 'tester'}
+    assert result == {'scopes': ['assume:anonymous', 'test:a', 'test:b'], 'clientId': 'tester'}
 
 
 def test_unicode_permacred_simple():
@@ -806,7 +806,7 @@ def test_unicode_permacred_simple():
         'clientScopes': ['test:a'],
         'requiredScopes': ['test:a'],
     })
-    assert result == {'scopes': ['test:a'], 'clientId': 'tester'}
+    assert result == {'scopes': ['assume:anonymous', 'test:a'], 'clientId': 'tester'}
 
 
 def test_invalid_unicode_permacred_simple():
@@ -856,7 +856,7 @@ def test_temporary_credentials():
         'clientScopes': ['test:*'],
         'requiredScopes': ['test:xyz'],
     })
-    assert result == {'scopes': ['test:xyz'], 'clientId': 'tester'}
+    assert result == {'scopes': ['assume:anonymous', 'test:xyz'], 'clientId': 'tester'}
 
 
 def test_named_temporary_credentials():
@@ -877,7 +877,7 @@ def test_named_temporary_credentials():
         'clientScopes': ['test:*', 'auth:create-client:credName'],
         'requiredScopes': ['test:xyz'],
     })
-    assert result == {'scopes': ['test:xyz'], 'clientId': 'credName'}
+    assert result == {'scopes': ['assume:anonymous', 'test:xyz'], 'clientId': 'credName'}
 
 
 def test_temporary_credentials_authorizedScopes():
@@ -898,7 +898,7 @@ def test_temporary_credentials_authorizedScopes():
         'clientScopes': ['test:*'],
         'requiredScopes': ['test:xyz:abc'],
     })
-    assert result == {'scopes': ['test:xyz:abc'], 'clientId': 'tester'}
+    assert result == {'scopes': ['assume:anonymous', 'test:xyz:abc'], 'clientId': 'tester'}
 
 
 def test_named_temporary_credentials_authorizedScopes():
@@ -920,7 +920,7 @@ def test_named_temporary_credentials_authorizedScopes():
         'clientScopes': ['test:*', 'auth:create-client:credName'],
         'requiredScopes': ['test:xyz:abc'],
     })
-    assert result == {'scopes': ['test:xyz:abc'], 'clientId': 'credName'}
+    assert result == {'scopes': ['assume:anonymous', 'test:xyz:abc'], 'clientId': 'credName'}
 
 
 def test_signed_url():
@@ -938,7 +938,7 @@ def test_signed_url():
     response = response.json()
     response['scopes'].sort()
     assert response == {
-        'scopes': sorted(['test:*', u'auth:create-client:test:*']),
+        'scopes': sorted(['assume:anonymous', 'test:*', u'auth:create-client:test:*']),
         'clientId': 'tester',
     }
 
@@ -975,7 +975,7 @@ def test_temp_credentials_signed_url():
     response.raise_for_status()
     response = response.json()
     assert response == {
-        'scopes': ['test:*'],
+        'scopes': ['assume:anonymous', 'test:*'],
         'clientId': 'tester',
     }
 
@@ -994,7 +994,7 @@ def test_signed_url_authorizedScopes():
     response.raise_for_status()
     response = response.json()
     assert response == {
-        'scopes': ['test:authenticate-get'],
+        'scopes': ['assume:anonymous', 'test:authenticate-get'],
         'clientId': 'tester',
     }
 
@@ -1017,6 +1017,6 @@ def test_temp_credentials_signed_url_authorizedScopes():
     response.raise_for_status()
     response = response.json()
     assert response == {
-        'scopes': ['test:authenticate-get'],
+        'scopes': ['assume:anonymous', 'test:authenticate-get'],
         'clientId': 'tester',
     }

--- a/clients/client-web/test/Auth_test.js
+++ b/clients/client-web/test/Auth_test.js
@@ -63,7 +63,7 @@ describe('Auth', function() {
       .buildSignedUrl(auth.testAuthenticateGet)
       .then(url => request(url))
       .then(({ scopes }) => {
-        expect(scopes).to.deep.equal(['test:authenticate-get', 'test:foo']);
+        expect(scopes).to.deep.equal(['assume:anonymous', 'test:authenticate-get', 'test:foo']);
       });
   });
 
@@ -83,7 +83,7 @@ describe('Auth', function() {
       .buildSignedUrl(auth.testAuthenticateGet)
       .then(url => request(url))
       .then(({ scopes }) => {
-        expect(scopes).to.deep.equal(['test:authenticate-get', 'test:bar']);
+        expect(scopes).to.deep.equal(['assume:anonymous', 'test:authenticate-get', 'test:bar']);
       });
   });
 
@@ -121,7 +121,7 @@ describe('Auth', function() {
       .buildSignedUrl(auth.testAuthenticateGet)
       .then(url => request(url))
       .then(({ scopes }) => {
-        expect(scopes).to.deep.equal(['test:authenticate-get']);
+        expect(scopes).to.deep.equal(['assume:anonymous', 'test:authenticate-get']);
       });
   });
 
@@ -186,17 +186,13 @@ describe('Auth', function() {
       })
       .then(({ clientId, scopes }) => {
         expect(clientId).to.equal('tester');
-        expect(scopes).to.deep.equal([]);
+        expect(scopes).to.deep.equal(['assume:anonymous']);
       });
   });
 
   it('should request with authentication and query string', () => {
     const auth = new Auth({
       rootUrl: helper.rootUrl,
-      credentials: {
-        clientId: 'tester',
-        accessToken: 'no-secret',
-      },
     });
 
     return auth
@@ -221,7 +217,7 @@ describe('Auth', function() {
       })
       .then(({ clientId, scopes }) => {
         expect(clientId).to.equal('tester');
-        expect(scopes).to.deep.equal(['test:param']);
+        expect(scopes).to.deep.equal(['assume:anonymous', 'test:param']);
       });
   });
 
@@ -290,7 +286,7 @@ describe('Auth', function() {
       })
       .then(({ clientId, scopes }) => {
         expect(clientId).to.equal('tester');
-        expect(scopes).to.deep.equal(['scopes:specific']);
+        expect(scopes).to.deep.equal(['assume:anonymous', 'scopes:specific']);
       });
   });
 
@@ -370,7 +366,7 @@ describe('Auth', function() {
       })
       .then(({ clientId, scopes }) => {
         expect(clientId).to.equal('my-temp-cred');
-        expect(scopes).to.deep.equal(['scopes:specific']);
+        expect(scopes).to.deep.equal(['assume:anonymous', 'scopes:specific']);
       });
   });
 
@@ -400,7 +396,7 @@ describe('Auth', function() {
       })
       .then(({ clientId, scopes }) => {
         expect(clientId).to.equal('my-temp-cred');
-        expect(scopes).to.deep.equal(['scopes:specific', 'scopes:another']);
+        expect(scopes).to.deep.equal(['assume:anonymous', 'scopes:specific', 'scopes:another']);
       });
   });
 
@@ -425,7 +421,7 @@ describe('Auth', function() {
       })
       .then(({ clientId, scopes }) => {
         expect(clientId).to.equal('tester');
-        expect(scopes).to.deep.equal(['scopes:subcategory:specific']);
+        expect(scopes).to.deep.equal(['assume:anonymous', 'scopes:subcategory:specific']);
       });
   });
 

--- a/clients/client/test/creds_test.js
+++ b/clients/client/test/creds_test.js
@@ -35,7 +35,7 @@ suite(testing.suiteName(), function() {
         requiredScopes: [],
       }), {
         clientId: 'tester',
-        scopes: [],
+        scopes: ['assume:anonymous'],
       });
   });
 
@@ -64,7 +64,7 @@ suite(testing.suiteName(), function() {
         requiredScopes: ['scopes:specific'],
       }), {
         clientId: 'tester',
-        scopes: ['scopes:specific'],
+        scopes: ['assume:anonymous', 'scopes:specific'],
       });
   });
 
@@ -92,7 +92,7 @@ suite(testing.suiteName(), function() {
         requiredScopes: ['scopes:specific'],
       }), {
         clientId: 'tester',
-        scopes: ['scopes:specific'],
+        scopes: ['assume:anonymous', 'scopes:specific'],
       });
   });
 
@@ -144,7 +144,7 @@ suite(testing.suiteName(), function() {
         requiredScopes: ['scopes:specific'],
       }), {
         clientId: 'my-temp-cred',
-        scopes: ['scopes:specific'],
+        scopes: ['assume:anonymous', 'scopes:specific'],
       });
   });
 
@@ -166,7 +166,7 @@ suite(testing.suiteName(), function() {
         requiredScopes: ['scopes:subcategory:specific'],
       }), {
         clientId: 'tester',
-        scopes: ['scopes:subcategory:specific'],
+        scopes: ['assume:anonymous', 'scopes:subcategory:specific'],
       });
   });
 
@@ -227,7 +227,7 @@ suite(testing.suiteName(), function() {
         requiredScopes: ['scopes:specific'],
       }), {
         clientId: 'my-temp-cred',
-        scopes: ['scopes:specific', 'scopes:another'],
+        scopes: ['assume:anonymous', 'scopes:specific', 'scopes:another'],
       });
   });
 
@@ -252,7 +252,7 @@ suite(testing.suiteName(), function() {
     });
     let url = cl.buildSignedUrl(cl.testAuthenticateGet);
     assert.deepEqual((await getJson(url)).scopes,
-      ['test:authenticate-get', 'test:foo']);
+      ['assume:anonymous', 'test:authenticate-get', 'test:foo']);
   });
 
   test('buildSignedUrl authorizedScopes (unauthorized)', async () => {
@@ -283,7 +283,7 @@ suite(testing.suiteName(), function() {
     });
     let url = cl.buildSignedUrl(cl.testAuthenticateGet);
     assert.deepEqual((await getJson(url)).scopes,
-      ['test:authenticate-get', 'test:bar']);
+      ['assume:anonymous', 'test:authenticate-get', 'test:bar']);
   });
 
   test('buildSignedUrl with temporary credentials and expiration', async () => {
@@ -338,7 +338,7 @@ suite(testing.suiteName(), function() {
       authorizedScopes: ['test:authenticate-get'],
     });
     let url = cl.buildSignedUrl(cl.testAuthenticateGet);
-    assert.deepEqual((await getJson(url)).scopes, ['test:authenticate-get']);
+    assert.deepEqual((await getJson(url)).scopes, ['assume:anonymous', 'test:authenticate-get']);
   });
 
   suite('Get with credentials from environment variables', async () => {
@@ -383,7 +383,7 @@ suite(testing.suiteName(), function() {
           requiredScopes: [],
         }), {
           clientId: 'tester',
-          scopes: [],
+          scopes: ['assume:anonymous'],
         });
     });
   });

--- a/workers/generic-worker/tcproxy/tcproxy_test.go
+++ b/workers/generic-worker/tcproxy/tcproxy_test.go
@@ -9,6 +9,7 @@ import (
 
 	tcclient "github.com/taskcluster/taskcluster/v38/clients/client-go"
 	"github.com/taskcluster/taskcluster/v38/clients/client-go/tcauth"
+	"github.com/taskcluster/taskcluster/v38/internal/scopes"
 	"github.com/taskcluster/taskcluster/v38/internal/testrooturl"
 )
 
@@ -53,7 +54,11 @@ func TestTcProxy(t *testing.T) {
 	if err != nil {
 		t.Fatalf("Could not interpret response %q as json: %v", string(data), err)
 	}
-	if len(scopeset.Scopes) != 1 || scopeset.Scopes[0] != "queue:get-artifact:SampleArtifacts/_/X.txt" {
-		t.Fatalf("Got incorrect data: %v", string(data))
+
+	// check that the current scopes satisfy the authorized scopes
+	given := scopes.Given(scopeset.Scopes)
+	required := scopes.Required([][]string{[]string{"queue:get-artifact:SampleArtifacts/_/X.txt"}})
+	if ok, err := given.Satisfies(required, tcauth.New(nil, rootURL)); !ok || err != nil {
+		t.Fatalf("Got current scopes that do not %s: %v", required, string(data))
 	}
 }

